test: fuzz non atomic bridging

0xteddybear opened the pull request.

0xteddybear committed:
test: changed mocked messenger ABI for message sending but kept assertions the same

0xteddybear committed:
docs: add new properties 26&27

0xteddybear committed:
test: queue cross-chain messages and test related properties

0xteddybear committed:
test: relay random messages from queue and check associated invariants

0xteddybear committed:
chore: rename bridge->senderc20 method for consistency with relayerc20

0xteddybear committed:
test: not-yet-deployed supertokens can get funds sent to them

0xteddybear committed:
chore: medusa runs forever by default

doable since it also handles SIGINTs gracefully

simon-something commented on packages/contracts-bedrock/test/properties/helpers/MockL2ToL2CrossDomainMessenger.t.sol:
```diff
@@ -37,22 +55,63 @@ contract MockL2ToL2CrossDomainMessenger {
+    }
+
+    function relayMessageFromQueue(uint256 index) public {
+        CrossChainMessage memory message = _messageQueue[index];
```
`index = index % _messageQueue.length - 1` ?

simon-something reviewed (approved):
noooice

0xteddybear committed:
chore: document the reason behind relay zero and send zero inconsistencies

0xteddybear committed:
fix: feedback from doc

0xteddybear committed:
fix: walk around possible medusa issue

I'm getting an 'unknown opcode 0x4e' in ProtocolAtomic constructor when
calling the MockL2ToL2CrossDomainMessenger for the first time

0xteddybear committed:
test: unguided handler for sendERC20

0xDiscotech commented on packages/contracts-bedrock/test/properties/medusa/properties/Protocol.t.sol:
```diff
@@ -27,18 +27,40 @@ contract ProtocolProperties is ProtocolGuided, ProtocolUnguided, CryticERC20Exte
         // iterate over unique deploy salts aka supertokens that are supposed to be compatible with each other
-        for (uint256 deploySaltIndex; deploySaltIndex < ghost_totalSupplyAcrossChains.length(); deploySaltIndex++) {
-            uint256 totalSupply;
+        for (uint256 deploySaltIndex = 0; deploySaltIndex < ghost_totalSupplyAcrossChains.length(); deploySaltIndex++) {
+            uint256 totalSupply = 0;
```
```suggestion
        for (uint256 deploySaltIndex; deploySaltIndex < ghost_totalSupplyAcrossChains.length(); deploySaltIndex++) {
            uint256 totalSupply;
```

0xDiscotech commented on packages/contracts-bedrock/test/properties/medusa/properties/Protocol.t.sol:
```diff
@@ -27,18 +27,40 @@ contract ProtocolProperties is ProtocolGuided, ProtocolUnguided, CryticERC20Exte
+    function property_totalSupplyAcrossChainsEqualsMintsWhenQueueIsEmpty() external view {
+        require(MESSENGER.messageQueueLength() == 0);
+        // iterate over unique deploy salts aka supertokens that are supposed to be compatible with each other
+        for (uint256 deploySaltIndex = 0; deploySaltIndex < ghost_totalSupplyAcrossChains.length(); deploySaltIndex++) {
+            uint256 totalSupply = 0;
```
```suggestion
        for (uint256 deploySaltIndex; deploySaltIndex < ghost_totalSupplyAcrossChains.length(); deploySaltIndex++) {
            uint256 totalSupply;
```

0xDiscotech commented on packages/contracts-bedrock/test/properties/helpers/MockL2ToL2CrossDomainMessenger.t.sol:
```diff
@@ -37,22 +55,64 @@ contract MockL2ToL2CrossDomainMessenger {
+    function relayMessageFromQueue(uint256 rawIndex) public {
+        uint256 index = rawIndex % _messageQueue.length;
+        CrossChainMessage memory message = _messageQueue[index];
+        _messageQueue[index] = _messageQueue[_messageQueue.length - 1];
```
Missing check `_length > 0` here so it doesn't revert due an underflow when subtracting 1?

0xDiscotech replied:
Actually, I think that check should go on the first line, since if `0`, the modulo operation will revert

0xteddybear replied:
medusa (and foundry, most of the time) ignore reverts in invariant tests, and only consider `assert(false)`, aka `Panic(0x01)`, as test failures. That's why we `try` and `catch` on function calls where we want to fail the test if it reverts: because otherwise it is ignored.

coverage reports are then used to know if tests are reverting too early and leaving assertions or state transitions uncovered. There are many other places throughout the campaign where execution might return early due to a revert, and many places where we explicitly revert instead of returning because it's easier :D

0xDiscotech replied:
Oh amazing, good to know :)

0xDiscotech commented on packages/contracts-bedrock/test/properties/medusa/fuzz/Protocol.guided.t.sol:
```diff
@@ -73,14 +75,86 @@ contract ProtocolGuided is ProtocolHandler {
+            );
+        } catch {
+            // 7
+            assert(false);
```
couldn't it fail when the length of relayed messages is `0` and then we need to check that if it failed was for that reason?

0xteddybear replied:
in that case the revert will be internal to this contract and not easy to `try/catch`, but we can safely ignore it in that case and medusa will re-try more often the call combinations where relays happen after queueing messages, since that leads to more coverage

0xDiscotech commented on packages/contracts-bedrock/test/properties/medusa/fuzz/Protocol.guided.t.sol:
```diff
@@ -73,14 +75,86 @@ contract ProtocolGuided is ProtocolHandler {
+    /// @custom:property relayERC20 increases sender's balance in the destination chain exactly by the input amount
+    /// @custom:property-id 7
+    /// @custom:property calls to relayERC20 always succeed as long as the cross-domain caller is valid
+    function fuzz_RelayERC20(uint256 messageIndex) external {
```
nit
```suggestion
    function fuzz_RelayERC20(uint256 messageIndex) external {
```

0xDiscotech reviewed (commented):
The way you handled the complexity is truly amazing.
Just left few small comments and a couple question, but looking very good so far!

0xteddybear committed:
fix: feedback from disco
